File: Behavioral/Lab22/C#/TemplateMethods/AddRoomTemplate.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Input;
using C_.Entities;

namespace C_.TemplateMethods
{
    public class AddRoomTemplate : AbstractTemplate
    {
        public AddRoomTemplate(params string[] args) : base(args)
        {
        }

        protected override string AddEmpty()
        {
            var agency = Agency.GetInstance();
            agency.AddEstate(new Room());
            return "Added empty new Room";
        }

        protected override string AddEstate()
        {
            if(_params.Length != 5) 
                return ReturnError();
            var agency = Agency.GetInstance();
            agency.AddEstate(new Room(_params[0], _params[1], Convert.ToDouble(_params[2]), Convert.ToDouble(_params[3]), _params[4]));
            return "Added new Room with args: " + string.Join(", ", _params);
        }

        // protected override string ReturnError()
        // {
        //     return "Failed to add new Room with args: " + string.Join(", ", _params);
        // }
    }
}

File: Behavioral/Lab22/Typescript/src/TemplateMethods/add_room_template.ts
import { Agency } from "../Entities/agency";
import { Apartment } from "../Entities/apartment";
import { Room } from "../Entities/room";
import { ICommand } from "../Interfaces/command";
import { AbstractTemplate } from "./abstract_template";

export class AddRoomStrategy extends AbstractTemplate {
    constructor(...params: string[]) {
        super(...params);
    }
    protected override addEmpty(): string {
        const agency = Agency.GetInstance();
        agency.AddEstate(new Room());
        return "Added empty new Room";
    }

    protected override addEstate(): string {
        const agency = Agency.GetInstance();
        if (this.Params.length === 5) {
            agency.AddEstate(new Room(
                this.Params[0],
                this.Params[1],
                +this.Params[2],
                +this.Params[3],
                this.Params[4]
            ));
            return "Added new Room with args: " + this.Params.join(", ");
        }
        return this.returnError();
    }
}